src/libteam/patch/0015-add-support-for-custom-retry.patch

```diff
@@ -0,0 +1,505 @@
++	if (version == 0xf1) {
++		lacpdu->v2.actor_retry_tlv_type = 0x80;
++		lacpdu->v2.actor_retry_tlv_len = 0x04;
++		lacpdu->v2.partner_retry_tlv_type = 0x81;
++		lacpdu->v2.partner_retry_tlv_len = 0x04;
++	}
+ }
+ 
+ static bool lacpdu_check(struct lacpdu *lacpdu)
+@@ -100,14 +123,31 @@ static bool lacpdu_check(struct lacpdu *lacpdu)
+ 	/*
+ 	 * According to 43.4.12 version_number, tlv_type and reserved fields
+ 	 * should not be checked.
++	 *
++	 * However, as part of 802.1ax, the version number is used to indicate
++	 * whether there may be additional TLVs present or not, so it does
++	 * need to be checked.
+ 	 */
+ 
+-	if (lacpdu->subtype		!= 0x01 ||
+-	    lacpdu->actor_info_len	!= 0x14 ||
+-	    lacpdu->partner_info_len	!= 0x14 ||
+-	    lacpdu->collector_info_len	!= 0x10 ||
+-	    lacpdu->terminator_info_len	!= 0x00)
++	if (lacpdu->subtype	!= 0x01)
++		return false;
++	if (lacpdu->version_number == 0x01) {
++	    if (lacpdu->actor_info_len	!= 0x14 ||
++			lacpdu->partner_info_len	!= 0x14 ||
++			lacpdu->collector_info_len	!= 0x10 ||
++			lacpdu->v1.terminator_info_len	!= 0x00)
++			return false;
++	} else if (lacpdu->version_number == 0xf1) {
++	    if (lacpdu->actor_info_len	!= 0x14 ||
++			lacpdu->partner_info_len	!= 0x14 ||
++			lacpdu->collector_info_len	!= 0x10 ||
++			lacpdu->v2.actor_retry_tlv_len != 0x04 ||
++			lacpdu->v2.partner_retry_tlv_len != 0x04 ||
++			lacpdu->v2.terminator_info_len	!= 0x00)
++			return false;
++	} else {
+ 		return false;
++	}
+ 	return true;
+ }
+ 
+@@ -154,6 +194,10 @@ struct lacp {
+ #define		LACP_CFG_DFLT_MIN_PORTS_MAX 1024
+ 		enum lacp_agg_select_policy agg_select_policy;
+ #define		LACP_CFG_DFLT_AGG_SELECT_POLICY LACP_AGG_SELECT_LACP_PRIO
++		bool enable_retry_count;
++#define		LACP_CFG_DFLT_ENABLE_RETRY_COUNT true
++		uint8_t retry_count;
++#define		LACP_CFG_DFLT_RETRY_COUNT 3
+ 	} cfg;
+ 	struct {
+ 		bool carrier_up;
+@@ -185,6 +229,7 @@ struct lacp_port {
+ 	struct lacpdu_info actor;
+ 	struct lacpdu_info partner;
+ 	struct lacpdu_info __partner_last; /* last state before update */
++	int partner_retry_count;
+ 	bool periodic_on;
+ 	struct lacp_port *agg_lead; /* leading port of aggregator.
+ 				     * NULL in case this port is not selected */
+@@ -513,6 +558,22 @@ static int lacp_load_config(struct teamd_context *ctx, struct lacp *lacp)
+ 	}
+ 	teamd_log_dbg(ctx, "Using agg_select_policy \"%s\".",
+ 		      lacp_get_agg_select_policy_name(lacp));
++
++	err = teamd_config_bool_get(ctx, &lacp->cfg.enable_retry_count, "$.runner.enable_retry_count");
++	if (err)
++		lacp->cfg.enable_retry_count = LACP_CFG_DFLT_ENABLE_RETRY_COUNT;
++	teamd_log_dbg(ctx, "Using enable_retry_count \"%d\".", lacp->cfg.enable_retry_count);
++
++	err = teamd_config_int_get(ctx, &tmp, "$.runner.retry_count");
++	if (err) {
++		lacp->cfg.retry_count = LACP_CFG_DFLT_RETRY_COUNT;
++	} else if (tmp < 3) {
++		teamd_log_err("\"retry_count\" value is out of its limits.");
++		return -EINVAL;
```

Comment:
We could probably generate the error message here and continue with retry count of LACP_CFG_DFLT_RETRY_COUNT, and proceed with LACP_CFG_DFLT_RETRY_COUNT?

Reply:
Updated.